4221 fix explorer calls

L03TJ3 opened the pull request:
# Description

Several errors and failed api calls were reported. also it seemed the API was called extensively which should be roughly every 30 seconds, not accounting for when someone scrolls over the feed. that triggers the sync again.

One bug was found (cause of it not found yet) where a BRIDGE_IN transaction is wrongly created on the feed with regular txtype receive. this results in it thinking its a regular contract transaction and tries to fetch the contract name for it as a result
- [x] - for these items an additional isBridgeReceive check is added when using getCounterParty

Additional fixes
- [x] - counterPartySmallAvatar does not seem to be used, resulting in getCounterParty always being called for gd-contracts
- [x] - API can return a valid 200 response with an error message about 'limit reached'. so verify that the events from get logs is an array 
- [x] - we call getOTPLEvents for withdraw/cancel tx's simultaneously, possible resulting in to many requests p/s (likely when someone scrolls over the feed)


About # (link your issue here)
#4221

L03TJ3 commented:
@johnsmith-gooddollar @sirpy I added some fixes. 
The one I don't know how to handle is how to throttle these two calls properly: https://github.com/GoodDollar/GoodDAPP/blob/530198a94020ab302b1e678f11326c0f53bdc922/src/lib/wallet/GoodWalletClass.js#L480

In theory it should run every 30 seconds. but I believe the sync retriggers when scrolling and it loads the not yet loaded feed-items

sirpy commented on src/lib/userStorage/UserStorageClass.js:
```diff
@@ -786,7 +786,7 @@ export class UserStorage {
       FeedItemType.EVENT_TYPE_SENDBRIDGE,
     ]
 
-    if (counterPartyEvents.includes(type) && (!counterPartyFullName || !counterPartySmallAvatar)) {
```
that doesnt seem correct. the small avatar is used in feed as the image.
I understand that TXs that are not p2p won't have a small avatar, so we probably need that type of filter

sirpy replied:
let make sure we have some kind of retry for the regular getCounterParty

L03TJ3 commented on src/lib/API/api.js:
```diff
@@ -530,6 +533,10 @@ export class APIService {
+      if (!isArray(events)) {
+        return
+      }
+
```
1. added it to getAbi > used by walletconnectclient which already has a catch flow and log.error
2. getsourcecode seems to not need it since it returns res?.ContractName so would just result in undefined. (do we still need an error handler here?)
3. altered the safeguard for getLogs to only add the events which were successfully fetched

sirpy replied:
@L03TJ3 you are not addressing my points
1. decision between returning null, throwing exception or events collected until error based on how caller uses results, retry requirements or lastBlock
2. we have multiple places where we call APIs, tatum, celo, fuse

L03TJ3 commented:
@sirpy right, after the last couple of fixes it does not look like it triggers additional calls to sync when scrolling the feed

johnsmith-gooddollar commented on src/lib/API/api.js:
```diff
@@ -505,7 +510,10 @@ export class APIService {
 
-    const sender32 = `0x${sender.toLowerCase().slice(2).padStart(64, '0')}`
+    const sender32 = `0x${sender
+      .toLowerCase()
```
there should be utils at web3 like normalise address, padValue/padBytes etc

L03TJ3 replied:
so I have to create an instance of web3 in the API class, just to add padLeft?
not sure I understand what this solves?

johnsmith-gooddollar replied:
import { padLeft } from 'web3-utils'

L03TJ3 replied:
done

sirpy commented on src/lib/API/api.js:
```diff
@@ -531,7 +539,10 @@ export class APIService {
       params.page += 1
-      txs.push(...events)
+
+      if (isArray(events)) {
```
still not correct.
if not array need special handling. see next lines that do events.length assuming it is an array.
also it means there's an error, maybe rate limit? does the code that uses it will retry? relies on running in intervals?
all of that needs to be verified and checked that our logic stays correct.

did you verify the lastBlock check I mentioned?

L03TJ3 replied:
1. all api explorer calls now have a array check
2. I believe to understand in the flow now it is handling a 'retry' based on the final line of syncTxFromExplorer:
```
    const lastBlock = Number(last(results)?.blockNumber)
    return lastBlock ? lastBlock + 1 : startBlock
```
since it resets to the original start-block when no blockNumber can be determined so will retry based on the polling interval set

3. see comment:

sirpy commented on src/lib/userStorage/FeedStorage.js:
```diff
@@ -564,9 +564,15 @@ export class FeedStorage {
     const isBridgeOrBuy = [TxType.TX_BRIDGE_IN, TxType.TX_BRIDGE_OUT, TxType.TX_BUYGD].includes(feedEvent.txType)
 
-    if (isContract !== '0x' && !isBridgeOrBuy) {
+    if (isContract !== '0x' && !isBridgeOrBuy && !isBridgeReceive) {
```
what is the difference between isBridgeReceive and TX_BRIDGE_IN ? isnt it supposed to be the same, seems like a double check to me

L03TJ3 replied:
Through various channels I explain there is a bug in the creation of bridge receive feed items, which are wrongfully created as regular receive (So not as TX_BRIDGE_IN).

This is a hotfix, but I will spend the time now on fixing the bug that causes it

L03TJ3 replied:
Okay. so after debugging it seems this is only related to older tx-items in the feed which are defined as regular TX_RECEIVE types.
For new feed-items it is set as expected: TX_BRIDGE_IN

Since we also decided to only run this when creating the items, this extra check does not seem to be needed anymore.
Removed

sirpy commented on src/lib/userStorage/UserStorageClass.js:
```diff
@@ -750,7 +750,7 @@ export class UserStorage {
   async formatEvent(event: FeedEvent) {
     try {
-      return await this._formatEvent(event)
+      return this._formatEvent(event)
```
the await is required for the catch block to happen

L03TJ3 replied:
reverted

L03TJ3 commented on src/lib/wallet/GoodWalletClass.js:
```diff
@@ -477,19 +477,30 @@ export class GoodWallet {
+        results?.length > 0 ? this.processEvents(results, startBlock) : undefined,
+        otpResults?.length > 0 ? this.processEvents(flatten(otpResults), startBlock) : undefined,
+      ])
+    } catch (error) {
```
@johnsmith-gooddollar @sirpy 
I now added a try/catch block here to log an error for the isArray safe-guard.
having this try/catch here means it will still finalize the syncTxFromExplorer and update lastblock accordingly.

Should we already log here? or do we push it all to the outer-catch block of syncTxFromExplorer caller and run the flow for syncTxFromBlockchain?

L03TJ3 replied:
2. Aren't they both catched because of the Promise.all on line 488? Where we await the resolution of both tokenPromise/otpPromise?

sirpy replied:
Promise all executes them simultaenously therefore more chances to hit rate limit. if you do
await getOTPL
await getOTPL
there's less chance to hit rate limits.

L03TJ3 replied:
I understand that, I was referring specifically to your question 'where is the exception handling for getTransactionsFromexplorer' and I say: its in the
```
Promise.all([tokenPromise <-- getTxFromEpxlorer, otpPromise <-- getOTPevents])
```

johnsmith-gooddollar replied:
```javascript
let withdrawEvents = []
let cancelEvents = []

const getOTPL = hash => API.getOTPLEvents(account, networkId, otpAddress, startBlock, currentBlock, hash)

const tokenEvents = await API.getTokenTxs(tokenAddress, account, networkId, startBlock).then(results =>
  results.map(result => ({ ...result, transactionHash: result.transactionHash || result.hash })),
)

await this.processEvents(tokenEvents)

if (otpAddress) {
  withdrawEvents = await getOTPL(withdrawHash)
  cancelEvents = await getOTPL(cancelHash)
}

// will parallel it all together, see fn implementation
await this.processEvents([
   ...tokenEvents,
  ...withdrawEvents,
  ...cancelEvents, 
])
```

johnsmith-gooddollar commented on src/lib/API/api.js:
```diff
@@ -531,7 +540,14 @@ export class APIService {
+
+      if (isArray(events)) {
+        txs.push(...events)
+      } else {
```
1. do not need, just if (!isArray(events)) { throw error }
2. you could just log.warn { events, params, chainId } here, do not add data if it's only needs for logging. in FV we used errors with custom .name props because it was used in the FV flow

johnsmith-gooddollar commented on src/lib/wallet/GoodWalletClass.js:
```diff
@@ -477,19 +477,30 @@ export class GoodWallet {
+      // so we log the error and let the sync finish to update lastBlock (potentially fetched some pages)
+      // if no data it is an unexpected exception so we let the outer catch handle it
+      if (error.data) {
+        log.error('syncTxFromExplorer failed', error.message, error.data)
```
do not this this is really needed, see my first comment

L03TJ3 commented:
@johnsmith-gooddollar @sirpy 
Latest changes are:
1. throw error when results is not the expected array without data, that is only being logged as warning
2. removed usage of Promise.all with the given code from @johnsmith-gooddollar 
3. removed inner try-catch from syncTxFromExplorer, this so the outer try-catch is used and it will when exception is thrown for explorer retry fetching with syncTxFromBlockchain

sirpy commented on src/lib/wallet/GoodWalletClass.js:
```diff
@@ -465,44 +465,27 @@ export class GoodWallet {
-    const lastBlock = Number(last(results)?.blockNumber)
+    await this.processEvents([...txEvents, ...withdrawEvents, ...cancelEvents])
+
+    const lastBlock = Number(last(txEvents)?.blockNumber)
```
this should be the max blockNumber from all events
